packages/gatsby/src/schema/node-model.js

```diff
@@ -0,0 +1,178 @@
+    ConnectionArgs
+  ): any | null;
+  getNodesByIds(
+    { ids: Array<IDOrNode>, type?: TypeOrTypeName },
+    ConnectionArgs
+  ): Array<any>;
+  getAllNodes({ type?: TypeOrTypeName }, ConnectionArgs): Array<any>;
+  runQuery(args: any): Promise<any>;
+  getTypes(): Array<string>;
+}
+
+class LocalNodeModel {
+  constructor({ schema, nodeStore, createPageDependency }) {
+    this.schema = schema
+    this.nodeStore = nodeStore
+    this.createPageDependency = createPageDependency
+  }
+
+  getNodeById({ id, type }, pageDependencies) {
+    const node = getNodeById(this.nodeStore, id)
+    let result
+    if (!node) {
+      result = null
+    } else if (!type) {
+      result = node
+    } else {
+      const nodeTypeNames = toNodeTypeNames(this.schema, type)
+      result = nodeTypeNames.includes(node.internal.type) ? node : null
+    }
+    return trackPageDependencies(
+      result,
+      pageDependencies,
+      this.createPageDependency
+    )
+  }
+
+  getNodesByIds({ ids, type }, pageDependencies) {
+    const nodes = ids.map(id => getNodeById(this.nodeStore, id)).filter(Boolean)
+    let result
+    if (!type) {
+      result = nodes
+    } else {
+      const nodeTypeNames = toNodeTypeNames(this.schema, type)
+      result = nodes.filter(node => nodeTypeNames.includes(node.internal.type))
+    }
+    return trackPageDependencies(
+      result,
+      pageDependencies,
+      this.createPageDependency
+    )
+  }
+
+  getAllNodes({ type }, pageDependencies) {
+    let result
+    if (!type) {
+      result = this.nodeStore.getNodes()
+    } else {
+      const nodeTypeNames = toNodeTypeNames(this.schema, type)
+      const nodes = nodeTypeNames.reduce(
+        (acc, typeName) => acc.concat(this.nodeStore.getNodesByType(typeName)),
+        []
+      )
+      result = nodes.filter(Boolean)
+    }
+
+    return trackPageDependencies(
+      result,
+      pageDependencies,
+      this.createPageDependency
+    )
+  }
+
+  async runQuery(args, pageDependencies) {
+    const { query, firstOnly, type } = args || {}
+
+    let nodes
+    const nodeTypeNames = toNodeTypeNames(this.schema, type)
+    if (nodeTypeNames.length > 1) {
+      nodes = nodeTypeNames.reduce(
+        (acc, typeName) => acc.concat(this.nodeStore.getNodesByType(typeName)),
+        []
+      )
+    }
+
+    const queryResult = await this.nodeStore.runQuery({
+      queryArgs: query,
+      firstOnly,
+      gqlType: typeof type === `string` ? this.schema.getType(type) : type,
+      nodes,
+    })
+    let result = queryResult
+    if (args.firstOnly) {
+      if (result && result.length > 0) {
+        result = result[0]
+      } else {
+        result = null
+      }
+    }
+    return trackPageDependencies(
+      result,
+      pageDependencies,
+      this.createPageDependency
+    )
+  }
+
+  getTypes() {
+    return this.nodeStore.getTypes()
+  }
+
+  findRootNodeAncestor(...args) {
+    return this.nodeStore.findRootNodeAncestor(...args)
```

Comment:
Is `findRootNodeAncestor` on the nodeStore now?

Reply:
Good find.